File: sdks/fog/challenges/create_server.rb
#!/usr/bin/env ruby

require 'fog'
Excon.defaults[:ssl_verify_peer] = false

service = Fog::Compute.new({
    :provider             => 'rackspace',
    :rackspace_username   => ENV['RAX_USERNAME'],
    :rackspace_api_key    => ENV['RAX_API_KEY'],
    :rackspace_region     => ENV['RAX_REGION'].downcase.to_sym,
    :rackspace_auth_url   => "#{ENV['RAX_AUTH_URL']}/v2.0"
})

server = service.servers.create :name => 'Fog Server',
                       :flavor_id => ENV['SERVER1_FLAVOR'].to_i,
                       :image_id => ENV['SERVER1_IMAGE']

puts "\n"

begin
  # Check every 5 seconds to see if server is in the active state (ready?). 
  # If the server has not been built in 5 minutes (600 seconds) an exception will be raised.
  server.wait_for(600, 5) do
    print "."
    STDOUT.flush
    ready?
  end
  
  puts "[DONE]\n\n"
  
rescue Fog::Errors::TimeoutError
  puts "[TIMEOUT]\n\n"
  
  puts "This server is currently #{server.progress}% into the build process and is taking longer to complete than expected."
  puts "You can continute to monitor the build process through the web console at https://mycloud.rackspace.com/\n\n" 
end

File: sdks/php-opencloud/challenges/create_server.php
<?php
require('vendor/autoload.php');
use OpenCloud\Rackspace;

$endpoint = getenv('RAX_AUTH_URL') . '/v2.0/';
$credentials = array(
    'username' => getenv('RAX_USERNAME'),
    'apiKey'   => getenv('RAX_API_KEY')
);

$rackspace = new Rackspace($endpoint, $credentials);
$compute   = $rackspace->computeService('cloudServersOpenStack', getenv('RAX_REGION'));

$image_id  = getenv('SERVER1_IMAGE');
$flavor_id = getenv('SERVER1_FLAVOR');

// Create a server in DFW
$server    = $compute->Server();
// create it
print("Creating server...");
$server->create(array(
    'name' => 'php-opencloud server',
    // Using the image ID from ORD
    'image' => $compute->image($image_id),
    // And a flavor that's too small
    'flavor' => $compute->flavor($flavor_id)
));
print("requested, now waiting...\n");
print("ID=" . $server->id . "...\n");
$server->WaitFor("ACTIVE", 600, 'dot');
print("done\n");
exit(0);

function dot($server)
{
    printf("%s %3d%%\n", $server->status, $server->progress);
}

?>
